4-bit QLoRA via bitsandbytes (4-bit base model + LoRA)

TimDettmers opened the pull request:
# What does this PR do?

This PR introduces 4-bit QLoRA to transformers. The main changes are for the bitsandbytes config. Additionally, we added one change to the LLaMA implementation where there is a bug that the data type chances if layer norms are in 32-bit and the rest is in bf16.

More information about QLoRA from our abstract:
>We develop QLoRA tuning, a method that finetunes by backpropagating gradients through a frozen 4-bit base model into low rank adapters (LoRA). With QLoRA tuning we can finetune 30B/65B parameter models on 24/48GB GPUs while preserving regular 16-bit full finetuning runtime and task performance. We achieve the memory efficiency and quantization precision through a combination of new methods: nested quantization to reduce the average memory footprint from 4.5 to 4.1 bits per parameter, paged optimizers to manage gradient checkpointing memory spikes, and a new data type, 4-bit NormalFloat (NF4), which is information theoretically and empirically optimal for normally distributed weights. To demonstrate the effectiveness and ease of use of QLoRA tuning we finetune more than 1,000 models to create a detailed dissection of instruction following performance across datasets (FLAN, Alpaca, Chip2, SuperNatural Instructions, Chip2, AnthropicHH), models types (LLaMA, T5), and model scales (125M to 65B). A discussion of the results is forthcoming in our paper.


## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

@sgugger @younesbelkada @sourabh112

TimDettmers committed:
Added lion and paged optimizers and made original tests pass.

TimDettmers committed:
Added lion and paged optimizers and made original tests pass.

TimDettmers committed:
Added lion and paged optimizers and made original tests pass.

TimDettmers committed:
Added tests for paged and lion optimizers.

TimDettmers committed:
Added tests for paged and lion optimizers.

TimDettmers committed:
Added tests for paged and lion optimizers.

TimDettmers committed:
Added and fixed optimizer tests.

TimDettmers committed:
Added and fixed optimizer tests.

TimDettmers committed:
Added and fixed optimizer tests.

TimDettmers committed:
Style and quality checks.

TimDettmers committed:
Style and quality checks.

TimDettmers committed:
Style and quality checks.

TimDettmers committed:
Initial draft. Some tests fail.

TimDettmers committed:
Initial draft. Some tests fail.

TimDettmers committed:
Initial draft. Some tests fail.

TimDettmers committed:
Merge remote-tracking branch 'origin/main' into bnb_4bit

TimDettmers committed:
Fixed dtype bug.

TimDettmers committed:
Fixed dtype bug.

TimDettmers committed:
Fixed dtype bug.

TimDettmers committed:
Fixed bug caused by torch_dtype='auto'.

TimDettmers committed:
Fixed bug caused by torch_dtype='auto'.

TimDettmers committed:
Fixed bug caused by torch_dtype='auto'.

TimDettmers committed:
All test green for 8-bit and 4-bit layers.

TimDettmers committed:
All test green for 8-bit and 4-bit layers.

TimDettmers committed:
All test green for 8-bit and 4-bit layers.

TimDettmers committed:
Merge remote-tracking branch 'forked/bnb_paged_optimizers' into forked/bnb_beta

TimDettmers committed:
Added fix for fp32 layer norms and bf16 compute in LLaMA.

TimDettmers committed:
Added fix for fp32 layer norms and bf16 compute in LLaMA.

TimDettmers committed:
Merge remote-tracking branch 'origin/main' into bnb_beta

younesbelkada commented:
Amazing! 
cc @SunMarc for visibility as well! 
Will review asap 💪

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_23479). All of your documentation changes will be reflected on that endpoint.

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -1700,6 +1705,15 @@ def save_pretrained(
             )
 
+        if getattr(self, "is_loaded_in_4bit", False):
+            warnings.warn(
+                "You are calling `save_pretrained` to a 8-bit converted model you may likely encounter unexepected"
+                " behaviors. If you want to save 8-bit models, make sure to have `bitsandbytes>0.37.2` installed.",
+                UserWarning,
+            )
+            raise NotImplementedError("You are calling `save_pretrained` on a 4-bit converted model. \
+                                      This is currently not supported")
```
No need for a warning if we throw an error.

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -2215,30 +2231,25 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
                     " bitsandbytes `pip install -i https://test.pypi.org/simple/ bitsandbytes` or"
                     " pip install bitsandbytes` "
                 )
-            if torch_dtype != torch.float16:
-                # We force the `dtype` to be float16, this is a requirement from `bitsandbytes`
-                logger.warning(
-                    f"Overriding torch_dtype={torch_dtype} with `torch_dtype=torch.float16` due to "
-                    "requirements of `bitsandbytes` to enable model loading in mixed int8. "
-                    "Either pass torch_dtype=torch.float16 or don't pass this argument at all to remove this warning."
-                )
-                torch_dtype = torch.float16
```
As in the Accelerate PR, is this not necessary anymore in the current released version?

TimDettmers replied:
It was never needed in the first place. The problem with this is that it does not allow for other data types which are necessary for mixed-precision 4-bit training.

younesbelkada replied:
Added a proper warning message!

sgugger commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -81,14 +81,12 @@ def __init__(self, hidden_size, eps=1e-6):
         hidden_states = hidden_states * torch.rsqrt(variance + self.variance_epsilon)
 
-        # convert into half-precision if necessary
-        if self.weight.dtype in [torch.float16, torch.bfloat16]:
-            hidden_states = hidden_states.to(self.weight.dtype)
-
-        return self.weight * hidden_states
+        hidden_states = (self.weight * hidden_states).to(input_dtype)
+        return hidden_states
```
The Llama modeling file should not be touched by this PR. If this change is necessary, it should go in its own PR.

sgugger commented on src/transformers/trainer.py:
```diff
@@ -1170,6 +1170,38 @@ def get_optimizer_cls_and_kwargs(args: TrainingArguments) -> Tuple[Any, Any]:
                 optimizer_kwargs.update(adam_kwargs)
             except ImportError:
                 raise ValueError("Trainer tried to instantiate apex FusedAdam but apex is not installed!")
+        elif args.optim in [
```
The optimizer should go in their own PRs, let's keep this one focused on one thing.

TimDettmers replied:
These changes would also be needed. Can you please have a look at: https://github.com/huggingface/transformers/pull/23217

sgugger commented on src/transformers/training_args.py:
```diff
@@ -139,10 +139,17 @@ class OptimizerNames(ExplicitEnum):
     ADAMW_ANYPRECISION = "adamw_anyprecision"
     SGD = "sgd"
     ADAGRAD = "adagrad"
+    ADAMW_BNB = "adamw_bnb_8bit"
```
Same here, this is part of something else.

sgugger commented on src/transformers/utils/bitsandbytes.py:
```diff
@@ -128,26 +135,43 @@ def replace_8bit_linear(model, threshold=6.0, modules_to_not_convert=None, curre
-
-        if len(list(module.children())) > 0:
-            replace_8bit_linear(module, threshold, modules_to_not_convert, current_key_name)
+        #current_key_name.append(name)
```
Should be removed if unnecessary.

sgugger commented on src/transformers/utils/quantization_config.py:
```diff
@@ -49,7 +53,7 @@ class BitsAndBytesConfig:
             magnitude ~5, but beyond that, there is a significant performance penalty. A good default threshold is 6,
             but a lower threshold might be needed for more unstable models (small models, fine-tuning).
-        llm_int8_skip_modules (`List[str]`, *optional*):
+        bnb_kbit_skip_modules (`List[str]`, *optional*):
```
We can't rename an argument here without care for backward compatibility. There are now quantization configs on the Hub with this argument set.

sgugger commented on tests/mixed_int8/test_mixed_int8.py:
```diff
@@ -1,689 +0,0 @@
-# coding=utf-8
```
We can't remove a whole part of the test suite like this.

TimDettmers replied:
I changed the name of the original test and added a bnb_kbit folder in tests. Added the missing files.

sgugger reviewed (commented):
Thanks for working on this. This PR contains three different changes that should all be in their own PR:
- the fixes in the LLama modeling file
- the optimizer from bitsandbytes
- the actual int4 quantization.

Moreover, Transformers has a very strong commitment to backward compatibility and while the quantization API is still experimental, we can't rename an argument in the quantization config without care as it will fail with existing quantization configs on the Hub. There must be some logic to accept the old name and change it to the new one added.

TimDettmers committed:
Fixing issues for PR #23479.

TimDettmers committed:
Merge branch 'bnb_beta' of github.com:timdettmers/transformers into bnb_beta

TimDettmers committed:
Reverted variable name change.

TimDettmers committed:
Added missing tests.

TimDettmers committed:
Merge branch 'bnb_beta' of github.com:timdettmers/transformers into forked/bnb_beta

TimDettmers committed:
Fixup changes.

TimDettmers committed:
Added fixup changes.

TimDettmers committed:
Merge branch 'bnb_beta' of github.com:timdettmers/transformers into bnb_beta

TimDettmers committed:
Missed some variables to rename.

TimDettmers commented:
Thank you for the review! I created a separate PR for the LLaMA bug: https://github.com/huggingface/transformers/pull/23535

The optimizers are also needed for the 4-bit fine-tuning of LLaMA 30B/65B on a single 24/48 GB GPU: https://github.com/huggingface/transformers/pull/23217

I reverted the names and removed code into other PRs. I added the missing test files. These, however are also renamed. Is the filename important for these?

Let me know if you see any other issues. Thank you, Sylvain!

younesbelkada commented on .github/workflows/build-nightly-ci-docker-images.yml:
I think this change and the changes below are not needed, probably a bad rebase happened

younesbelkada commented on src/transformers/modeling_utils.py:
```diff
@@ -2671,15 +2682,6 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
 
         if isinstance(device_map, str):
             special_dtypes = {}
-            if load_in_8bit:
-                special_dtypes.update(
-                    {
-                        name: torch_dtype
-                        for name, _ in model.named_parameters()
-                        if any(m in name for m in modules_to_not_convert)
-                    }
-                )
-
```
I think this block is needed as it will be used to keep the modules that are inside `modules_to_not_convert` in the desired `dtype` - also we might need to change the condition to `if load_in_8bit or load_in_4_bit`

younesbelkada replied:
Done!

younesbelkada commented on src/transformers/modeling_utils.py:
```diff
@@ -2796,10 +2798,13 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
 
+        model.is_loaded_in_4bit = load_in_4bit
         model.is_loaded_in_8bit = load_in_8bit
+        model.is_loaded_in_kbit = load_in_8bit or load_in_4bit
```
Should we create an intermediate variable  above `is_loaded_in_kbit = load_in_8bit or load_in_4bit` and use all the way down?

younesbelkada commented on src/transformers/trainer.py:
```diff
@@ -392,7 +392,7 @@ def __init__(
 
         # At this stage the model is already loaded
         if getattr(model, "is_loaded_in_8bit", False):
```
```suggestion
        if getattr(model, "is_loaded_in_kbit", False):
```

younesbelkada commented on src/transformers/utils/bitsandbytes.py:
```diff
@@ -52,12 +52,16 @@ class `Int8Params` from `bitsandbytes`.
+        is_4bit = False
     else:
-        has_fp16_weights = getattr(module._parameters[tensor_name], "has_fp16_weights", None)
+        is_4bit = isinstance(module._parameters[tensor_name], bnb.nn.Params4bit)
```
We should probably protect the import of `Params4bit` for users that uses a previous version of `bnb`

younesbelkada commented on src/transformers/utils/bitsandbytes.py:
```diff
@@ -75,11 +79,17 @@ class `Int8Params` from `bitsandbytes`.
+            if is_8bit:
+                new_value = bnb.nn.Int8Params(new_value, requires_grad=False, **kwargs).to(device)
+            elif is_4bit:
+                new_value = bnb.nn.Params4bit(new_value, requires_grad=False, **kwargs).to(device)
```
Same comment as above

younesbelkada commented on src/transformers/utils/quantization_config.py:
```diff
@@ -86,10 +116,41 @@ def post_init(self):
+            raise ValueError("bnb_4bit_quant_type must be a string")
+
+        if not isinstance(self.bnb_4bit_use_double_quant, bool):
+            raise ValueError("bnb_4bit_use_double_quant must be a boolean")
```
Or maybe the check of the appropriate bnb version can be done here!

younesbelkada commented on tests/repo_utils/test_get_test_info.py:
I think if you merge you current branch with upstream main this diff should disapear

younesbelkada commented on src/transformers/modeling_utils.py:
```diff
@@ -2720,7 +2722,7 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
             model.tie_weights()
             device_map = infer_auto_device_map(model, dtype=torch_dtype if not load_in_8bit else torch.int8, **kwargs)
```
Here I would definitely add a check to make sure  `device_map="auto"` would work as expected. One should import the recent `CustomDtype` from `accelerate` here: https://github.com/huggingface/accelerate/pull/1434 
and the logic flow would look like:
1- check if `accelerate` is on the appropriate version and import `CustomDtype` and pass `CustomDtype.int4`
2- if not, log a warning and suggest users to upgrade accelerate
Wdyt?

```python
if load_in_4_bit:
    _accelerate_has_custom_dtype = version.parse(importlib_metadata.version("accelerate")) > version.parse("0.19.0")
    if _accelerate_has_custom_dtype:
        from accelerate.utils import CustomDtype
        target_dtype = CustomDtype.int4
    else:
       logger.warn("You are using `device_map="auto"` on a 4bit loaded version of the model. To compute the appropriate device map, you should upgrade your `accelerate` library, `pip install --upgrade accelerare` or install it from source to support int4 auto device map calculation. You may encounter unexpected behavior, or pass your own device map")
       target_dtype = torch.int8


```

younesbelkada commented on tests/bnb_kbit/test_mixed_int8.py:
```diff
@@ -77,11 +77,11 @@ class BaseMixedInt8Test(unittest.TestCase):
     # Constant values
     EXPECTED_RELATIVE_DIFFERENCE = (
-        1.540025  # This was obtained on a Quadro RTX 8000 so the number might slightly change
+        1.054829776346287  # This was obtained on a Quadro RTX 8000 so the number might slightly change
```
Hmm I think this shouldn't be changed, you probably tested on another device than a T4?

younesbelkada replied:
Ok after some investigaation it seems to be related to https://github.com/huggingface/transformers/pull/23479/files#r1200368398
Could you please revert this change, and just change the way `self.model_8bit` is loaded in `setUp` to load it with `torch_dtype=torch.float16` ?

younesbelkada commented on tests/bnb_kbit/test_mixed_int8.py:
```diff
@@ -77,11 +77,11 @@ class BaseMixedInt8Test(unittest.TestCase):
 
     input_text = "Hello my name is"
-    EXPECTED_OUTPUT = "Hello my name is John.\nI am a friend of the family.\n"
+    EXPECTED_OUTPUT = "Hello my name is John.\nI am a friend of your father.\n"
```
Same here maybe it's because you ran your test on a device different than a T4?

younesbelkada commented on tests/trainer/test_trainer.py:
These changes should be there as well ! 💪

younesbelkada reviewed (commented):
Amazing and inspiring work @TimDettmers !
I left few minor comments - I also think merging your branch with upstream main branch should remove the unneeded diffs
My major comment being the fix regarding `device_map="auto"` issues users could encounter, otherwise the integration looks very smooth!
I would also question ourselves whether we need to protect bnb import when importing 4bit modules e.g.: https://github.com/TimDettmers/bitsandbytes/issues/416

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into HEAD

younesbelkada committed:
revert trainer tests

younesbelkada committed:
revert test trainer

younesbelkada committed:
another revert

younesbelkada committed:
fix tests and safety checkers

younesbelkada committed:
protect import

younesbelkada committed:
simplify a bit

younesbelkada committed:
Update src/transformers/trainer.py

younesbelkada committed:
few fixes

younesbelkada committed:
add warning

younesbelkada committed:
replace with `load_in_kbit = load_in_4bit or load_in_8bit`

younesbelkada committed:
fix test

younesbelkada committed:
fix tests

younesbelkada committed:
this time fix tests

younesbelkada committed:
safety checker

younesbelkada committed:
add docs

sgugger commented on docs/source/en/main_classes/quantization.mdx:
```diff
@@ -19,8 +19,43 @@ This is supported by most of the GPU hardwares since the `0.37.0` release of `bi
+
+#### Load a large model in 4bit
+
+You can load a model by roughly quartering the memory requirements by using `load_in_4bit=True` argument when calling `.from_pretrained` method
```
```suggestion
By using `load_in_4bit=True` when calling the `.from_pretrained` method, you can divide your memory use by 4 (roughly).
```

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -1700,6 +1701,12 @@ def save_pretrained(
             )
 
+        if getattr(self, "is_loaded_in_4bit", False):
+            raise NotImplementedError(
+                "You are calling `save_pretrained` on a 4-bit converted model. \
+                                      This is currently not supported"
+            )
```
Format is wrong here
```suggestion
            raise NotImplementedError(
                "You are calling `save_pretrained` on a 4-bit converted model. This is currently not supported"
            )
```

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -1876,29 +1883,29 @@ def get_memory_footprint(self, return_buffers=True):
+        if getattr(self, "is_loaded_in_kbit", False):
             raise ValueError(
-                "`.to` is not supported for `8-bit` models. Please use the model as it is, since the"
+                "`.to` is not supported for `k-bit` models. Please use the model as it is, since the"
```
```suggestion
                "`.to` is not supported for `4-bit` or `8-bit` models. Please use the model as it is, since the"
```
The end user won't know what `k-bit` stands for.

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -1876,29 +1883,29 @@ def get_memory_footprint(self, return_buffers=True):
+        if getattr(self, "is_loaded_in_kbit", False):
             raise ValueError(
-                "`.half()` is not supported for `8-bit` models. Please use the model as it is, since the"
+                "`.half()` is not supported for `k-bit` models. Please use the model as it is, since the"
```
Same as above.

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -2688,6 +2702,25 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
+                    "0.19.0"
+                )
+
+                if _accelerate_has_custom_dtype:
```
Do the version test here.

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -2688,6 +2702,25 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
+                    from accelerate.utils import CustomDtype
+
+                    target_dtype = CustomDtype.INT4
+                else:
+                    logger.warning(
+                        "You are using `device_map='auto'` on a 4bit loaded version of the model. To automatically compute the appropriate device map, you should upgrade your `accelerate` library, `pip install --upgrade accelerare` or install it from source to support int4 auto device map calculation. You may encounter unexpected behavior, or pass your own device map"
+                    )
+                    target_dtype = torch.int8
```
Here the user has indicated `load_in_4bit=True` with improper version of Accelerate, so we should throw an error instead of defaulting to 8bit.

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -2795,11 +2828,13 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
 
+        model.is_loaded_in_4bit = load_in_4bit
         model.is_loaded_in_8bit = load_in_8bit
+        model.is_loaded_in_kbit = load_in_8bit or load_in_4bit
```
`is_loaded_in_kbit` is not a good name. All models are lodaded in a certain kbit precision for k in {4, 8, 16, 32}. `is_quantized` would be more adapted.

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -2862,12 +2897,12 @@ def _load_pretrained_model(
-        if load_in_8bit:
-            from .utils.bitsandbytes import set_module_8bit_tensor_to_device
+        if load_in_kbit:
+            from .utils.bitsandbytes import set_module_kbit_tensor_to_device
```
Same for this function.

sgugger commented on src/transformers/utils/bitsandbytes.py:
```diff
@@ -15,7 +15,7 @@
 
 
-def set_module_8bit_tensor_to_device(module, tensor_name, device, value=None, fp16_statistics=None):
+def set_module_kbit_tensor_to_device(module, tensor_name, device, value=None, fp16_statistics=None):
```
```suggestion
def set_module_quantized_tensor_to_device(module, tensor_name, device, value=None, fp16_statistics=None):
```

sgugger commented on src/transformers/utils/bitsandbytes.py:
```diff
@@ -15,7 +15,7 @@
     from accelerate.utils import find_tied_parameters
 
 
-def set_module_8bit_tensor_to_device(module, tensor_name, device, value=None, fp16_statistics=None):
```
This function can't be deleted just yet without a deprecation cycle.

sgugger commented on src/transformers/utils/bitsandbytes.py:
```diff
@@ -95,10 +105,10 @@ class `Int8Params` from `bitsandbytes`.
             module._parameters[tensor_name] = new_value
 
 
-def replace_8bit_linear(model, threshold=6.0, modules_to_not_convert=None, current_key_name=None):
```
Same for this one.

sgugger commented on tests/bnb_kbit/test_mixed_int8.py:
```diff
@@ -105,7 +105,9 @@ def setUp(self):
             self.model_name, torch_dtype=torch.float16, device_map="auto"
         )
-        self.model_8bit = AutoModelForCausalLM.from_pretrained(self.model_name, load_in_8bit=True, device_map="auto")
+        self.model_8bit = AutoModelForCausalLM.from_pretrained(
+            self.model_name, load_in_8bit=True, device_map="auto", torch_dtype=torch.float16
+        )
```
The fact we now need to add a `torch_dtype` is a very worrying breaking change.

sgugger reviewed (commented):
Thanks for the additional work on this. There is still some work needed:
- backward compatibility should be taken care of when renaming functions
- kbit is not a very good name. A torch.float32 is kbit with k=32. A name with quantization in it would make more sense.
- the current API for loaded in 8bit should not change for the user (I see a lot of `torh_dtype=torch.float16` added)
- the API should not ask the user to provide two different dtypes (`load_in_4bit=True` and `torch_dtype=torch.float16`)

younesbelkada committed:
revert torch_dtype

younesbelkada committed:
Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

younesbelkada committed:
multiple fixes

younesbelkada committed:
update docs

younesbelkada committed:
version checks and multiple fixes

younesbelkada committed:
replace `is_loaded_in_kbit`

younesbelkada committed:
replace `load_in_kbit`

younesbelkada committed:
change methods names

younesbelkada committed:
better checks

younesbelkada reviewed (commented):
Thanks for the extensive review @sgugger , I should have address all the comments now

younesbelkada committed:
oops

younesbelkada committed:
oops

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -2215,30 +2225,31 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
+
+            if torch_dtype is None:
                 # We force the `dtype` to be float16, this is a requirement from `bitsandbytes`
                 logger.warning(
```
Should be an info IMO here. The warning is only interesting for a power user.

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -2215,30 +2225,31 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
-                    "requirements of `bitsandbytes` to enable model loading in mixed int8. "
-                    "Either pass torch_dtype=torch.float16 or don't pass this argument at all to remove this warning."
+                    "requirements of `bitsandbytes` to enable model loading in mixed kbit. "
+                    "Pass your own torch_dtype to specify the dtype of the remaining non-linear layers or pass torch_dtype=torch.float16 to remove this warning."
```
Nit: make sure to respect the 119 char limit.

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -2688,6 +2714,20 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
+
+                    target_dtype = CustomDtype.INT4
+                else:
+                    raise ValueError(
+                        "You are using `device_map='auto'` on a 4bit loaded version of the model. To automatically compute the appropriate device map, you should upgrade your `accelerate` library, `pip install --upgrade accelerare` or install it from source to support fp4 auto device map calculation. You may encounter unexpected behavior, or pass your own device map"
+                    )
```
Same here, make sure to respect the 119 char limit.

sgugger commented on src/transformers/utils/bitsandbytes.py:
```diff
@@ -128,29 +135,61 @@ def replace_8bit_linear(model, threshold=6.0, modules_to_not_convert=None, curre
 
+# For backward compatibility
+def replace_8bit_linear(*args, **kwargs):
+    print("`replace_8bit_linear` will be deprecated in a future version, please use `replace_with_bnb_linear` instead")
```
Using `warnings.warn(xxx, FutureWarning)` here

younesbelkada replied:
oops sorry!

sgugger commented on src/transformers/utils/bitsandbytes.py:
```diff
@@ -128,29 +135,61 @@ def replace_8bit_linear(model, threshold=6.0, modules_to_not_convert=None, curre
+
+# For backward compatiblity
+def set_module_8bit_tensor_to_device(*args, **kwargs):
+    print(
```
Same here.

sgugger commented on tests/bnb_kbit/test_mixed_int8.py:
```diff
@@ -476,7 +494,7 @@ def test_pipeline(self):
             "text-generation",
             model=self.model_name,
-            model_kwargs={"device_map": "auto", "load_in_8bit": True},
+            model_kwargs={"device_map": "auto", "load_in_8bit": True, "torch_dtype": torch.float16},
```
Should be left as is.

sgugger reviewed (commented):
Thanks for iterating. I have a couple more comments. Can you also rename the test folder to `quantization` or `bitsandbytes` (but not kbit)?

sgugger reviewed (commented):
Thanks for iterating. I have a couple more comments. Can you also rename the test folder to `quantization` or `bitsandbytes` (but not kbit)?

younesbelkada committed:
address final comments

diegomontoya commented:
@TimDettmers  Can the bitsandbytes team provide a corresponding bitsandbytes branch that corresponds to this PR?  The "closed_beta" on bitsandbytes is the closest I found but it doesn't appear to be final/rc-quality and contains debug print logs such as 

https://github.com/TimDettmers/bitsandbytes/compare/main...closed_beta#diff-4d235c7e595546c6656c229dfa139298ce6602b356c2d0bafcb2352eb2cfae79R222

Without the proper branch/link to bitsandbytes changes, it is very hard to test this. Since this PR is public, the bnb branch should no longer be in closed beta. 

Thank you.

artidoro commented:
I think the changes for QLoRA were recently merged in the main branch of bitsandbytes.
